// @unique-nft/utils/exchangeInfo.mjs
import "./chunk-SMH5LAMQ.mjs";

// src/ExchangeInfo/prices.ts
var URLS = {
  QTZ: {
    mexc: "https://www.mexc.com/open/api/v2/market/ticker?symbol=QTZ_USDT",
    mexcCorsProxied: "https://api.allorigins.win/get?url=https%3A%2F%2Fwww.mexc.com%2Fopen%2Fapi%2Fv2%2Fmarket%2Fticker%3Fsymbol%3DQTZ_USDT"
  },
  UNQ: {
    huobi: "https://api.huobi.pro/market/detail/merged?symbol=unqusdt"
  }
};
var getQTZPrice = async (dontUseCORSProxy) => {
  const url = dontUseCORSProxy ? URLS.QTZ.mexc : URLS.QTZ.mexcCorsProxied;
  const response = await (await fetch(url)).json();
  const result = dontUseCORSProxy ? response.data[0] : JSON.parse(response.contents).data[0];
  const ask = parseFloat(result.ask);
  const bid = parseFloat(result.ask);
  const avg = parseFloat(((ask + bid) / 2).toFixed(5));
  return {
    coin: "QTZ",
    avg,
    currency: "USDT"
  };
};
var getUNQPrice = async () => {
  const result = (await (await fetch(URLS.UNQ.huobi)).json()).tick;
  const ask = result.ask[0];
  const bid = result.bid[0];
  const avg = parseFloat(((ask + bid) / 2).toFixed(5));
  return {
    coin: "UNQ",
    avg,
    currency: "USDT"
  };
};

// src/ExchangeInfo/index.ts
var ExchangeInfo = {
  getQTZPrice,
  getUNQPrice
};
var ExchangeInfo_default = ExchangeInfo;
export {
  ExchangeInfo,
  ExchangeInfo_default as default
};
