// @unique-nft/utils/exchangeInfo.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/ExchangeInfo/index.ts
var ExchangeInfo_exports = {};
__export(ExchangeInfo_exports, {
  ExchangeInfo: () => ExchangeInfo,
  default: () => ExchangeInfo_default
});
module.exports = __toCommonJS(ExchangeInfo_exports);

// src/ExchangeInfo/prices.ts
var URLS = {
  QTZ: {
    mexc: "https://www.mexc.com/open/api/v2/market/ticker?symbol=QTZ_USDT",
    mexcCorsProxied: "https://api.allorigins.win/get?url=https%3A%2F%2Fwww.mexc.com%2Fopen%2Fapi%2Fv2%2Fmarket%2Fticker%3Fsymbol%3DQTZ_USDT"
  },
  UNQ: {
    huobi: "https://api.huobi.pro/market/detail/merged?symbol=unqusdt"
  }
};
var getQTZPrice = async (dontUseCORSProxy) => {
  const url = dontUseCORSProxy ? URLS.QTZ.mexc : URLS.QTZ.mexcCorsProxied;
  const response = await (await fetch(url)).json();
  const result = dontUseCORSProxy ? response.data[0] : JSON.parse(response.contents).data[0];
  const ask = parseFloat(result.ask);
  const bid = parseFloat(result.ask);
  const avg = parseFloat(((ask + bid) / 2).toFixed(5));
  return {
    coin: "QTZ",
    avg,
    currency: "USDT"
  };
};
var getUNQPrice = async () => {
  const result = (await (await fetch(URLS.UNQ.huobi)).json()).tick;
  const ask = result.ask[0];
  const bid = result.bid[0];
  const avg = parseFloat(((ask + bid) / 2).toFixed(5));
  return {
    coin: "UNQ",
    avg,
    currency: "USDT"
  };
};

// src/ExchangeInfo/index.ts
var ExchangeInfo = {
  getQTZPrice,
  getUNQPrice
};
var ExchangeInfo_default = ExchangeInfo;
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ExchangeInfo
});
